build.gradle

```diff
@@ -1226,15 +1226,27 @@ project(':generator') {
 }
 
 project(':clients') {
+  apply plugin: 'com.github.johnrengelman.shadow'
+
   archivesBaseName = "kafka-clients"
 
+  shadowJar {
+    zip64 = true
+    relocate('io.opentelemetry', 'shadow.io.opentelemetry')
+    relocate('com.google.protobuf', 'shadow.com.google.protobuf')
+  }
+
   dependencies {
     implementation libs.zstd
     implementation libs.lz4
     implementation libs.opentelemetryProto
     implementation libs.snappy
     implementation libs.slf4jApi
```

Comment:
we need to make sure none of the existing libraries marked as dependencies are included in the fat jar, and that they continue to be declared as runtime dependencies in the final kafka-clients pom.
You may have to mark those dependencies as `shadow` (see https://imperceptiblethoughts.com/shadow/configuration/#configuring-the-runtime-classpath and what we do for the ce-metrics project) 
To verify you have the desired output:
1. Check the final clients jar file to make sure that there are no other classes in the jar, other than the ones defined by the clients project and the relocated proto / otel classes. Especially, there should not be any classes outside of the org.apache.kafka and the shaded package. Anything else means you have extraneous dependencies being included in there.
2. Check the final generated client pom file to make sure it still contains the same dependencies as it does today, e.g. compare it with https://search.maven.org/artifact/org.apache.kafka/kafka-clients/3.1.0/jar

Reply:
Ah I think I get it now - so we mark the dependencies we want to shadow as `shadow`, and the version of them will be defined in the `dependencies.gradle` file.  To avoid class conflict we relocate these jars to a separated directory.